keps/sig-api-machinery/20190721-explain-categories-expand.md

```diff
@@ -0,0 +1,198 @@
+---
+title: Explain Categories and Expand
+authors:
+  - "@feloy"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-cli
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-07-21
+last-updated: 2019-07-25
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Explain Categories and Expand
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Categorize the fields of PodSpec](#categorize-the-fields-of-podspec)
+    - [Expand some fields of Container](#expand-some-fields-of-container)
+    - [Order fields of VolumeMount](#order-fields-of-volumemount)
+- [Compatibility](#compatibility)
+<!-- /toc -->
+
+## Summary
+
+Actually, documentation (especially `kubectl explain`) describing Kinds of the Kubernetes API
+lists all the fields of a Kind in the alphabetic order, and there is not enough information in the declaration of these fields to list them in a specific order.
+
+We propose to add an annotation `x-kubernetes-explain-category` applicable to fields, that represents both
+the name of the category and the order of this category related to the other categories of the same Kind.
+
+In addition, we propose to add an annotation `x-kubernetes-explain-expand` applicable to fields, 
+that indicates if the field should be expanded in documentation in order to display inline the content 
+of this field.
+
+## Proposal
+
+### User Stories
+
+#### Categorize the fields of PodSpec
+
+To categorize a field, it is necessary to add a comment to the field in `types.go`:
+
+```
+// +k8s:openapi-gen=x-kubernetes-explain-category:1.category-name
+```
+
+The value (here *1.category-name*) contains two parts, separated by a dot (.).
```

Comment:
initial thoughts:
* This is too single-purpose. I'm OK with some sort of categorization tag, but "kubectl-explain" is too specific. Won't other tools be able to use this?
* This is doing two things at once. Keep the categories and their order separately configured. Why would the order need to be configured at all? kubectl can come up with a standardized order (and e.g. go alphabetical for categories it doesn't understand), or the order could be given separately in another extension tag.

Reply:
You may also want a place to put a longer-winded explanation/description of a category.